### PR TITLE
Implement exclusion and replacements for contributed bindings

### PR DESCRIPTION
**Pull request by vRallev**
This is the last PR for the contributing binding feature. It implements the various exclusion and replacement rules.

### CODE REVIEW

File: compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt [hunk 1]
```diff
@@ -400,6 +401,7 @@ class InterfaceMergerTest(
     }
   }
 
+  @Ignore("Need to investigate. Somehow these compilations are successful now.")
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/InterfaceMergerTest.kt:404]: Fixed in #53

File: compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt [hunk 2]
```diff
@@ -446,6 +535,7 @@ class ModuleMergerTest(
     }
   }
 
+  @Ignore("Need to investigate. Somehow these compilations are successful now.")
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/ModuleMergerTest.kt:538]: Fixed in #53

File: compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt [hunk 3]
```diff
@@ -141,6 +160,7 @@ internal class ModuleMerger(
         .map { it.first }
         .map { codegen.typeMapper.mapType(it) }
         .minus(replacedModules)
+        .minus(replacedModulesByContributedBindings)
```
Comment (kirillzh) [inline compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt:163]: Neat!

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt [hunk 4]
```diff
@@ -117,16 +142,41 @@ internal class BindingModuleGenerator(
...
+          .filter { it.annotationOrNull(daggerModuleFqName) != null }
+          .mapNotNull {
+            it.annotationOrNull(contributesToFqName, scope)
+                ?.replaces(module)
+          }
 
       val bindingMethods = (contributedBindingsThisModule + contributedBindingsDependencies)
+          .minus(replacedBindings)
+          .minus(bindingsReplacedInDaggerModules)
+          .minus(excludedTypesForScope[scope] ?: emptyList())
```
Comment (kirillzh) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt:179]: nit: `excludedTypesForScope[scope].orEmpty()` :P

**Pull request merged**